app/src/main/java/org/schabi/newpipe/player/mediabrowser/MediaBrowserConnector.java

```diff
@@ -0,0 +1,328 @@
+    }
+
+    private Single<List<MediaItem>> populatePlaylist(final long playlistId) {
+        final var playlist = getPlaylistManager().getPlaylistStreams(playlistId).firstOrError();
+        return playlist.map(items -> {
+            final List<MediaItem> results = new ArrayList<>();
+            int index = 0;
+            for (final var item : items) {
+                results.add(createPlaylistStreamMediaItem(playlistId, item, index));
+                ++index;
+            }
+            return results;
+        });
+    }
+
+    private void playbackError(@StringRes final int resId, final int code) {
+        playerService.stopForImmediateReusing();
+        sessionConnector.setCustomErrorMessage(playerService.getString(resId), code);
+    }
+
+    private void playbackError(@NonNull final ErrorInfo errorInfo) {
+        playbackError(errorInfo.getMessageStringId(), PlaybackStateCompat.ERROR_CODE_APP_ERROR);
+    }
+
+    private Single<PlayQueue> extractPlayQueueFromMediaId(final String mediaId) {
+        final Uri mediaIdUri = Uri.parse(mediaId);
+        if (mediaIdUri == null) {
+            return Single.error(new ContentNotAvailableException("Media ID cannot be parsed"));
+        }
+        if (mediaId.startsWith(ID_BOOKMARKS)) {
+            final var path = mediaIdUri.getPathSegments();
+            if (path.size() == 4) {
+                final long playlistId = Long.parseLong(path.get(2));
+                final int index = Integer.parseInt(path.get(3));
+
+                return getPlaylistManager()
+                        .getPlaylistStreams(playlistId)
+                        .firstOrError()
+                        .map(items -> {
+                            final var infoItems = items.stream()
+                                    .map(PlaylistStreamEntry::toStreamInfoItem)
+                                    .collect(Collectors.toList());
+                            return new SinglePlayQueue(infoItems, index);
+                        });
+            }
+        } else if (mediaId.startsWith(ID_STREAM)) {
+            final var path = mediaIdUri.getPathSegments();
+            if (path.size() == 3) {
+                final long streamId = Long.parseLong(path.get(2));
+                return getDatabase().streamHistoryDAO().getHistory()
+                        .firstOrError()
+                        .map(items -> {
+                            final var infoItems = items.stream()
+                                    .filter(it -> it.getStreamId() == streamId)
+                                    .map(StreamHistoryEntry::toStreamInfoItem)
+                                    .collect(Collectors.toList());
+                            return new SinglePlayQueue(infoItems, 0);
+                        });
+            }
+        }
+
+        return Single.error(new ContentNotAvailableException("Media ID cannot be parsed"));
+    }
```

Comment:
Extraction code is also part of Bookmarks/History Tab fragment classes. Can we put the code in a common util/handler class? That would make future refactoring much easier.

Reply:
This extraction code is for the media ID URIs used for interaction between the browser and the player, which I thought are not the same as the URIs NewPipe currently uses for streams. Did NewPipe already have this concept? 

I think it makes this code easier if these media IDs were used throughout the app, but it felt like an intrusive change to me.

Reply:
I think that's fine. What do you think @Stypox?